File: Algorithms/p131_Palindrome_Partitioning/p131_Palindrome_Partitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        if not s:
            return [[]]
        res = []
        cur = []
        self.partitionHelper(res,cur,s,0)
        return res

    def partitionHelper(self, res, cur, s, start):
        for i in range(start,len(s)):
            # check palindrome
            l = start
            r = i
            while l<r:
                if s[l]!=s[r]:
                    break
                l += 1
                r -= 1
            if l<r:
                continue
            # dfs search
            cur.append(s[start:i+1])
            if i<len(s)-1:
                self.partitionHelper(res,cur,s,i+1)
            else:
                res.append(cur[:])
            cur.pop()

File: Algorithms/p131_Palindrome_Partitioning/p131_Palindrome_Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        if(s.empty()) return {{}};
        vector<vector<string>> res;
        vector<string> cur;
        partitionHelper(res,cur,s,0);
        return res;
    }
    void partitionHelper(vector<vector<string>>& res, vector<string>& cur, string& s, int start) {
        for(int i=start;i<s.size();++i){
            // check palindrome
            int l = start;
            int r = i;
            for(;l<r;++l,--r){
                if(s[l]!=s[r]) break;
            }
            if(l<r) continue;
            // dfs search
            cur.push_back(s.substr(start,i-start+1));
            if(i<s.size()-1) partitionHelper(res,cur,s,i+1);
            else             res.push_back(cur);
            cur.pop_back();
        }
    }
};
